Marlin/src/HAL/HAL_STM32/HAL_timers_STM32.h

```diff
@@ -34,42 +34,66 @@
 #define FORCE_INLINE __attribute__((always_inline)) inline
 
 #define hal_timer_t uint32_t
-#define HAL_TIMER_TYPE_MAX 0xFFFF
+#define HAL_TIMER_TYPE_MAX 0xFFFFFFFF // Timers can be 16 or 32 bit
 
 #ifdef STM32F0xx
 
   #define HAL_TIMER_RATE (HAL_RCC_GetSysClockFreq()) // frequency of timer peripherals
 
-  #define STEP_TIMER 16
-  #define TEMP_TIMER 17
+  #ifndef STEP_TIMER
+    #define STEP_TIMER 16
+  #endif
+
+  #ifndef TEMP_TIMER
+    #define TEMP_TIMER 17
+  #endif
 
 #elif defined STM32F1xx
 
   #define HAL_TIMER_RATE (HAL_RCC_GetPCLK2Freq()) // frequency of timer peripherals
 
-  #define STEP_TIMER 4
-  #define TEMP_TIMER 2
+  #ifndef STEP_TIMER
+    #define STEP_TIMER 4
+  #endif
+
+  #ifndef TEMP_TIMER
+    #define TEMP_TIMER 2
+  #endif
 
 #elif defined STM32F4xx
 
-  #define HAL_TIMER_RATE (HAL_RCC_GetPCLK2Freq()) // frequency of timer peripherals
+  #define HAL_TIMER_RATE (F_CPU/2) // frequency of timer peripherals
```

Comment:
Actually there are three different ways. STM32F0xx uses `HAL_RCC_GetSysClockFreq()`....

I think the `F_CPU/2` method is better than calling `HAL_RCC_GetPCLK2Freq()` because it saves quite some CPU cycles. `F_CPU` will just read a global variable defined in the STM core. **BUT**, depending on clock configuration I think it may be that `F_CPU/2` not always equals `HAL_RCC_GetPLCK2Freq()` so it might be best to leave the code as is, e.g use `HAL_RCC_GetPCLK2Freq()`. I'll revert this change.

Reply:
I don't think saving some CPU cycles in this part of the code is that important, since the calculated values are only used for setting up the timers. They are not used in an interrupt or the main loop as far as I know.

The current assumption we take for the timer clocks are good educated guesses I think. If a board designer decided to use a weird timer clock, then he/she can still change that in the board definition.

Reply:
It is used in the stepper irq:

`hal_timer_t pulse_end = HAL_timer_get_count(PULSE_TIMER_NUM) + hal_timer_t(MIN_PULSE_TICKS);`
const hal_timer_t added_step_ticks = hal_timer_t(ADDED_STEP_TICKS);

`MIN_PULSE_TICKS` and `ADDED_STEP_TICKS` uses `PULSE_TIMER_PRESCALE` which in turn uses `HAL_TIMER_RATE`.